vpr/src/route/rr_node_storage.h

```diff
@@ -119,15 +130,177 @@ class t_rr_node_storage {
+    //      - edge_switch
+
+    /* Edge mutators */
+
+    // Reserve at least num_edges in the edge backing arrays.
+    void reserve_edges(size_t num_edges);
+
+    // Adds ones edge.  This method is efficient if reserve_edges was called with
+    // the number of edges present in the graph.  This method is still
+    // amortized O(1), like std::vector::emplace_back, but both runtime and
+    // peak memory usage will be higher if reallocation is required.
+    void emplace_back_edge(RRNodeId src, RRNodeId dest, short edge_switch);
+
+    // Adds a batch of edges.
+    void alloc_and_load_edges(const t_rr_edge_info_set* rr_edges_to_create);
+
+    /* Edge finalization methods */
+
+    // Counts the number of rr switches needed based on fan in.
+    //
+    // init_fan_in does not need to be invoked before this method.
+    size_t count_rr_switches(
+        size_t num_arch_switches,
+        t_arch_switch_inf* arch_switch_inf,
+        t_arch_switch_fanin& arch_switch_fanins) const;
+
+    // Maps arch_switch_inf indicies to rr_switch_inf indicies.
+    //
+    // This must be called before partition_edges if edges were created with
+    // arch_switch_inf indicies.
+    void remap_rr_node_switch_indices(const t_arch_switch_fanin& switch_fanin);
+
+    // Marks that edge switch values are rr switch indicies.
+    //
+    // This must be called before partition_edges if edges were created with
+    // rr_switch_inf indicies.
+    void mark_edges_as_rr_switch_ids();
+
+    // Sorts edge data such that configurable edges appears before
+    // non-configurable edges.
+    void partition_edges();
+
+    // Validate that edge data is partitioned correctly.
+    bool validate() const;
+
+    /* Edge accessors
+     *
+     * Only call these methods after partition_edges has been invoked. */
+    edge_idx_range edges(const RRNodeId& id) const {
+        return vtr::make_range(edge_idx_iterator(0), edge_idx_iterator(num_edges(id)));
+    }
+    edge_idx_range configurable_edges(const RRNodeId& id) const {
+        return vtr::make_range(edge_idx_iterator(0), edge_idx_iterator(num_edges(id) - num_non_configurable_edges(id)));
+    }
+    edge_idx_range non_configurable_edges(const RRNodeId& id) const {
+        return vtr::make_range(edge_idx_iterator(num_edges(id) - num_non_configurable_edges(id)), edge_idx_iterator(num_edges(id)));
+    }
+
+    t_edge_size num_edges(const RRNodeId& id) const {
+        RREdgeId first_id = first_edge_[id];
+        RREdgeId second_id = (&first_edge_[id])[1];
+        return (size_t)second_id - (size_t)first_id;
+    }
+
+    t_edge_size num_configurable_edges(const RRNodeId& id) const;
+    t_edge_size num_non_configurable_edges(const RRNodeId& id) const;
+
+    RREdgeId edge_id(const RRNodeId& id, t_edge_size iedge) const {
+        RREdgeId first_edge = first_edge_[id];
+        RREdgeId ret((size_t)first_edge + iedge);
+        VTR_ASSERT_SAFE(ret < (&first_edge_[id])[1]);
+        return ret;
+    }
```

Comment:
That we can do this easily given the edge ordering is a really nice feature, which I expect will make porting/refactoring downstream code much easier.